services/src/main/java/org/keycloak/protocol/AuthorizationEndpointBase.java

```diff
@@ -196,4 +215,49 @@ protected AuthenticationSessionModel createAuthenticationSession(ClientModel cli
 
     }
 
+    /**
+     * Checks if the request should be redirected to the default identity provider or not. This method returns {@code true}
+     * If a default identity provider is found (i.e. it was either explicitly set as the default provider for the realm or
+     * specified as the default provider for the auth request using the {@code kc_idp_hint} query parameter) and if this provider
+     * supports redirects with the {@code prompt=none} query parameter; it returns {@code false} otherwise.
+     *
+     * @param authSession a reference to the authentication session.
+     * @param protocol a reference to the login protocol.
+     * @return {@code true} if the request if {@code prompt=none} query parameter should be redirected to the default
+     * provider (if one exists); {@code false} otherwise.
+     */
+    private boolean redirectToDefaultProvider(final AuthenticationSessionModel authSession, final LoginProtocol protocol) {
+        if (protocol instanceof OIDCLoginProtocol) {
+
+            // check if "kc_idp_hint" query parameter has been set to set or override the default provider.
+            if (session.getContext().getUri().getQueryParameters().containsKey(AdapterConstants.KC_IDP_HINT)) {
+                String providerId = session.getContext().getUri().getQueryParameters().getFirst(AdapterConstants.KC_IDP_HINT);
+                if (providerId != null && !providerId.isEmpty()) {
+                    IdentityProviderModel provider = authSession.getRealm().getIdentityProviderByAlias(providerId);
+                    if (provider != null && provider.isEnabled() && Boolean.valueOf(provider.getConfig().get(SUPPORTS_PROMPT_NONE))) {
+                        return true;
+                    }
+                }
+                else {
+                    // "kc_idp_hint" was set with an empty value - cannot assume a default provider for the request.
+                    return false;
+                }
+            }
+
+            // check if we have a default identity provider configured in the realm.
+            List<AuthenticatorConfigModel> authConfigs = authSession.getRealm().getAuthenticatorConfigs();
+            if (authConfigs != null) {
+                for (AuthenticatorConfigModel authConfig : authConfigs) {
+                    if (authConfig.getConfig().containsKey(DEFAULT_PROVIDER)) {
+                        String providerId = authConfig.getConfig().get(DEFAULT_PROVIDER);
+                        if (providerId != null && !providerId.isEmpty()) {
+                            IdentityProviderModel provider = authSession.getRealm().getIdentityProviderByAlias(providerId);
+                            return provider != null && provider.isEnabled() && Boolean.valueOf(provider.getConfig().get(SUPPORTS_PROMPT_NONE));
+                        }
+                    }
+                }
+            }
+        }
+        return false;
+    }
```

Comment:
This code kinda duplicates to some extent the logic from the IdentityProviderAuthenticator to determine the default provider. An alternative would be to work on the Response that is returned when we call authenticateOnly() and see if that response contains a redirect (302) to the broker. We could then parse the Location URI to extract the provider alias and use it to check if the provider supports prompt=none redirects. In other words, if we get a 302 and URL matches the /broker/idp_alias/login structure we know it was assembled by the IdentityProviderAuthenticator and we can extract the idp_alias from the URL and use it.

I've attempted this alternative way but the resulting code looked a bit hacky to be honest with the URL manipulation. It is, however, the most efficient way to obtain the default IdP alias without going through the realm config or re-checking the kc_idp_hint param, so if you feel this is better than the code above I can switch to it.

Reply:
+1 that this code duplicates the login in IdentityProviderAuthenticator. I suggest another alternative, which won't require AuthorizationEndpointBase to be aware of any IDP specific logic - see description for more details.